pandas/_libs/ndframe_iter.c

```diff
@@ -0,0 +1,222 @@
+  if (managerLocs == NULL) {
+    return NULL;
+  }
+
+  // TODO: we could probably just supply managerLocs to the list()
+  // built-in instead of going from mgr_locs->as_array->tolist
+  ndarray = PyObject_GetAttrString(managerLocs, "as_array");
+  Py_DECREF(managerLocs);
+  if (ndarray == NULL) {
+    return NULL;
+  }
+
+  list = PyObject_CallMethod(ndarray, "tolist", NULL);
+  Py_DECREF(ndarray);
+  if (list == NULL) {
+    return NULL;
+  } else if (!PyList_Check(list)) {
+    Py_DECREF(list);
+    return NULL;
+  }
+  
+  return list;
+}
+
+
+// Given a DataFrame, this will create a struct containing both
+// the length of ndarrays the frame uses along with an array of
+// PyArrayObjects, appearing in C-order along the requested axis.
+//
+// For example, if we had a DataFrame that looks as follows
+//    a     b   c
+// 0  1   2.0   3
+// 1  4   5.0   6
+//
+// This would deconstruct the blocks and provide a struct
+// with len 3; accessing
+// ndarrays[0] will yield ndarray([1, 4])
+// ndarrays[1] will yield ndarray([2., 5.])
+// ndarrays[2] will yield ndarray([3, 6])
+//
+// The goal of this is to provide a performant way in C to
+// maintain axis order and dtype information of the
+// supplied DataFrame.
+PdOrderedArrays *PdOrderedArrays_New(PyObject *df, int axis) {
+  PyObject *blocks, *block, *managerLocs, *blockValues, *ndarr, *key;
+  PyObject **ndarrays;
+  Py_ssize_t i, j, loc, ncols;
+  PdOrderedArrays *result;
+
+  // 
+  assert(axis == 0);
+
+  ncols = getDimLength(df, 1);
+  blocks = getBlocksTuple(df);
+  if (blocks == NULL) {
+    return NULL;
+  }
+
+  ndarrays = PyObject_Malloc(sizeof(PyObject *) * ncols);
+  if (ndarrays == NULL) {
+    Py_DECREF(blocks);
+    return NULL;
+  }
+
+
+  // Iterate over all of the blocks, getting a slice from
+  // each block sequentially and storing it in the appropriate order
+  // in `arrays`
+  for (i = 0; i < PyTuple_GET_SIZE(blocks); i++) {
+    block = PyTuple_GET_ITEM(blocks, i);
+    
+    blockValues = PyObject_CallMethod(block, "get_block_values", NULL);
+    if (blockValues == NULL) {
+      Py_DECREF(blocks);
+      PyObject_Free(ndarrays);
+      return NULL;
+    }
+
+    managerLocs = getManagerLocationsAsList(block);
+    if (managerLocs == NULL) {
+      Py_DECREF(blockValues);
+      Py_DECREF(blocks);
+      PyObject_Free(ndarrays);
+      return NULL;
+    }
+
+    // managerLocs tells use where each column of the block
+    // exists in the maintaining dataframe, so iterate
+    // managerLocs sequentially and assign, get the column of
+    // data and assign that reference to the appropriate position
+    // in `ndarrays`
+    for (j = 0; j < PyList_GET_SIZE(managerLocs); j++) {
+      loc = PyLong_AsLongLong(PyList_GET_ITEM(managerLocs, j));
+      key = PyLong_FromLongLong(j);
+      if (key == NULL) {
+        goto LOOP_ERROR;
+      }
+
+      // TODO: We should actually be grabbing an ndarray reference here
```

Comment:
Since this is rather verbose, I should point out that this comment might be the most helpful in understanding what is going on here (or at least what should be)